internal/olm/client/status.go

```diff
@@ -71,13 +87,74 @@ func (c Client) GetObjectsStatus(ctx context.Context, objs ...runtime.Object) St
 	return Status{Resources: rss}
 }
 
-func (s Status) HasExistingResources() bool {
+// HasInstalledResources only returns true if at least one resource in s
+// was returned successfully by the API server. A resource error status
+// containing any error except "not found", or "no kind match" errors
+// for Custom Resources, will result in HasInstalledResources returning
+// false and the error.
+func (s Status) HasInstalledResources() (bool, error) {
+	crdKindSet, err := s.getCRDKindSet()
+	if err != nil {
+		return false, fmt.Errorf("error getting set of CRD kinds in resources: %w", err)
+	}
 	for _, r := range s.Resources {
 		if r.Resource != nil {
-			return true
+			return true, nil
+		} else if r.Error != nil && !apierrors.IsNotFound(r.Error) {
+			nkmerr := &meta.NoKindMatchError{}
+			if !errors.As(r.Error, &nkmerr) || !crdKindSet.Has(r.GVK.Kind) {
+				return false, r.Error
+			}
 		}
 	}
-	return false
+	return false, nil
+}
+
+// getCRDKindSet returns the set of all kinds specified by all CRDs in s.
+func (s Status) getCRDKindSet() (sets.String, error) {
+	crdKindSet := sets.NewString()
+	dec := serializer.NewCodecFactory(sch).UniversalDeserializer()
+	for _, r := range s.Resources {
+		if r.GVK.Kind == "CustomResourceDefinition" {
+			switch v := r.requestObject.(type) {
+			case *unstructured.Unstructured:
+				vb, err := v.MarshalJSON()
+				if err != nil {
+					return nil, err
+				}
+				// Use unversioned CustomResourceDefinition to avoid implementing cast
+				// for all versions.
+				obj, _, err := dec.Decode(vb, nil, nil)
+				if err != nil {
+					return nil, err
+				}
```

Comment:
Is it possible to use something like this: https://github.com/operator-framework/operator-sdk/blob/3c8e5e7280d8b8584e46605faba999e4fa3c4487/pkg/helm/internal/types/types.go#L138-L141